onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc

```diff
@@ -368,104 +393,171 @@ SubGraphCollection_t TensorrtExecutionProvider::GetSupportedList(SubGraphCollect
-    graph_build.AddInitializedTensor(*(tensor.second));
+    // Remove the node with empty input shape
+    if (!exclude_node) {
+      parser_nodes_vector.back().first.push_back(index);
+    } else if (!parser_nodes_vector.back().first.empty()) {
+      parser_nodes_vector.push_back({{}, false});
+    }
   }
 
-  auto status = graph_build.Resolve();
-  ORT_ENFORCE(status.IsOK(), status);
-  ONNX_NAMESPACE::ModelProto model_proto = model.ToProto();
-  model_proto.set_ir_version(ONNX_NAMESPACE::Version::IR_VERSION);
-
-  // Serialize modelproto to string
-  std::string string_buf;
-  model_proto.SerializeToString(&string_buf);
-
-  // Get supported node list
-  std::vector<size_t> nodes_vector(node_index.size());
-  std::iota(std::begin(nodes_vector), std::end(nodes_vector), 0);
-  SubGraphCollection_t parser_nodes_vector = {{nodes_vector, false}};
+  // Get supported node list from TensorRT parser
   SubGraphCollection_t supported_nodes_vector;
   bool early_termination = false;
   supported_nodes_vector = GetSupportedList(parser_nodes_vector, 0, max_partition_iterations_, graph, &early_termination);
   if (early_termination) {
     supported_nodes_vector.clear();
   }
 
-  // Remove nodes with empty shape (for example [1, 0]) because TensorRT 6 doens't support empty shape
-  SubGraphCollection_t post_processed_supported_nodes_vector;
-  for (auto& group : supported_nodes_vector) {
-    // Right now only NonZero and NonMaxSuppression related empty shape nodes are removed.
-    // The typical cases are Faster-rcnn and Mask-rcnn
-    // TODO: Remove the code if TensorRT fixed the issue in future release, or find a better generic way here to work around
-    post_processed_supported_nodes_vector.push_back({});
-    for (const auto& index : group.first) {
-      const auto& node = graph.GetNode(node_index[index]);
-      bool exclude_node = false;
-      for (auto input : node->InputDefs()) {
-        auto input_shape = input->Shape();
-        if (input_shape) {
-          for (auto dim : input_shape->dim()) {
-            std::string dim_name = dim.dim_param();
-            std::string exclude_dim_name1 = "NonZero";
-            std::string exclude_dim_name2 = "NonMaxSuppression";
-            if (!dim_name.empty()) {
-              if ((dim_name.find(exclude_dim_name1) != std::string::npos) || (dim_name.find(exclude_dim_name2) != std::string::npos)) {
-                exclude_node = true;
-                break;
-              }
-            }
+  // Remove subgraphs if its size is less than the predefined minimal size
+  for (auto it = supported_nodes_vector.begin(); it != supported_nodes_vector.end(); ++it) {
+    const int subgraph_size = it->first.size();
+    if (subgraph_size < min_subgraph_size_) {
+      supported_nodes_vector.erase(it--);
+    }
+  }
+
+  //Iteratively detect and remove cycles in the partitioned graph
+  std::vector<std::unique_ptr<ComputeCapability>> result;
+  bool trt_cycle = true;
+  while (trt_cycle) {
```

Comment:
GetCapability() function is getting too long.
can you break out this logic into separate function ?